### ISSUE TITLE
How to validate FormData object?

### ISSUE THREAD

**Issue by da1z** [colinhacks/zod#3333]
is there proper way to validate FormData type without it converting to object?

**Comment by colinhacks** [comment]
Not really. You can do `z.instanceof(FormData)` but that's about it. I've considered a `z.formdata()` API that would let you specify whether to `.get()` or `.getAll()` each field. Then you're left with `{[k: string]: string | string[] }`, and you'd have to use `.transform()` to do any additional coercion/transforms. Ultimately I wasn't sure it was worthwhile, but feel free to propose an API here,.

**Comment by LiamMartens** [comment]
I don't think this is useful since `FormData` objects are rarely validated like this imo. Coming into `express` they are just objects and when using forms in react or otherwise they are also generally stored as objects. So `FormData` is just a an object for transportation.

**Comment by da1z** [comment]
> I don't think this is useful since `FormData` objects are rarely validated like this imo. Coming into `express` they are just objects and when using forms in react or otherwise they are also generally stored as objects. So `FormData` is just a an object for transportation.

I would not say it is useless since form server action in next js receives FormData as parameter and it would be very useful to validate formdata with zod before processing it.

```typescript
export default function Page() {
  async function createInvoice(formData: FormData) {
    'use server'
 
    const rawFormData = {
      customerId: formData.get('customerId'),
      amount: formData.get('amount'),
      status: formData.get('status'),
    }
 
    // mutate data
    // revalidate cache
  }
 
  return <form action={createInvoice}>...</form>
}
```

**Comment by LiamMartens** [comment]
@da1z ah my bad! I didn't realize that's how next handled it 🤔 I wonder what this API would look like...

**Comment by da1z** [comment]
> @da1z ah my bad! I didn't realize that's how next handled it 🤔 I wonder what this API would look like...

i would prefer to keep existing api but make it compatible with formData object so I can use it like this:

```typescript
const schema = z.object({
  name: z.string(),
});
const parsed: z.infer<typeof schema> = schema.parse(formData);
```

**Comment by oom-** [comment]
@da1z I think this is the best way to go as well.

This would help for `Nuxt` too: 
```ts
const formDataSchema = z.object({
    id: z.number().min(0),
    avatarFile: z.instanceof(File)
});
 //Get params
const body = await readMultipartFormData(event, (formData) => formDataSchema .safeParse(formData));
```

**Comment by isaacgarc-dev** [comment]
But wouldn't you get id as a `string` type? I am facing the issue FormData sends number inputs as string How can i parse this without having to manually pick the attributes and cast them My zod schema fails since string is not a number

**Comment by jonathanhefner** [comment]
> I've considered a `z.formdata()` API that would let you specify whether to `.get()` or `.getAll()` each field. Then you're left with `{[k: string]: string | string[] }`, and you'd have to use `.transform()` to do any additional coercion/transforms. Ultimately I wasn't sure it was worthwhile, but feel free to propose an API here.

I think it was mentioned in another thread, but such an API could also work for `URLSearchParams`, which would be nice.

But for both `FormData` and `URLSearchParams`, there is an issue of interpreting nested keys.  For example, interpreting `foo.bar[0]` vs `foo[bar][0]`.

To handle all possible cases, what about a more abstract API:

  ```ts
  type Path = (string | number)[]
  type Value = any // or FormDataEntryValue for FormData specifically
  type PathValuePair = [Path, Value]

  let formData: FormData
  let formDataAdapter: (formData: FormData) => Generator<PathValuePair>

  let schema = z.object({
    /* ... */
  }).adapter(formDataAdapter)

  schema.parse(formData)
  ```

Where `formDataAdapter` is a function that accepts a `FormData` object and returns a series of path-value pairs.  zod could provide a few default adapters (e.g. `zod.adapters.formData`), but users could also specify their own.  Here is an example implementation:

  ```ts
  function* formDataAdapter(formData: FormData) {
    for (const [pathString, value] of formData.entries()) {
      const path: Path = pathString.split(/\.|(?=\[)/).map(key =>
        /^\[\d+\]/.test(key)
          ? parseInt(key.slice(1, -1), 10)
          : key
      )

      yield [path, value] as PathValuePair
    }
  }
  ```

When parsing, zod could pass the `FormData` to the adapter and build a parseable object from the output:

  ```ts
  function buildParseableObject(pairs: Generator<PathValuePair>) {
    const result = {}

    for (const [path, value] of pairs) {
      let current: any = result

      for (let i = 0; i < path.length - 1; i += 1) {
        const key = path[i]
        const nextKey = path[i + 1]
        current[key] ??= typeof nextKey === "number" ? [] : {}
        current = current[key]
      }

      current[path[path.length - 1]] = value
    }

    return result
  }
  ```

**Comment by da1z** [comment]
ok closing the issue. bot answer is good enough

**Comment by Kudze** [comment]
Fyi @da1z the bot answer is not good enough. Both zodix and remix-params-helper seem like no longer maintained. (last updates 2 years ago).

 The issue with `Object.fromEntries(formData.entries());` is that if form transfers arrays or objects. Then the zod validation will not work.

For example:
```
<input type="hidden" name="test[0][name]" value="test"/>
```

WIll produce:
```
{ "test[0][name]": "test"}
```

However the expected result would be:
```
{
  "test": [{"name": "test"}]
}
```

**Comment by da1z** [comment]
@Kudze I understand. I think one of the options would be to look into lodash.set method or plain js alternative and using along formData.entries()

**Comment by BlueManCZ** [comment]
I'd also recommend checking out [zod-form-data](https://github.com/airjp73/rvf/tree/main/packages/zod-form-data).

**Comment by shahrryyar** [comment]
Hey @colinhacks 👋 — I've been running into the FormData pain point constantly in Remix/React Router server actions, and noticed zod/v4 still doesn't have a first-class solution for it.

I've implemented **`z.formData()`** and opened a PR: it accepts a native `FormData` instance and returns a fully typed, validated plain object with automatic coercion:

- `z.number()` → `Number(value)`
- `z.boolean()` → checkbox semantics (`'on'/'true'/'1'` → `true`, absent key → `false`)
- `z.date()` → `new Date(value)`
- `z.array(element)` → `fd.getAll(key)` with per-element coercion
- `z.file()` → raw `File` passed through
- Everything else → string as-is

21 tests, 0 type errors, all 3,853 existing tests pass. Would love your feedback!
